content/doc/book/pipeline/cps-method-mismatches.adoc

```diff
@@ -0,0 +1,250 @@
+---
+layout: section
+title: Pipeline CPS Method Mismatches
+---
+ifdef::backend-html5[]
+:notitle:
+:description:
+:author:
+:email: jenkinsci-docs@googlegroups.com
+:sectanchors:
+ifdef::env-github[:imagesdir: ../resources]
+ifndef::env-github[:imagesdir: ../../resources]
+:hide-uri-scheme:
+:toc:
+endif::[]
+
+= Pipeline CPS Method Mismatches
+
+Jenkins Pipeline uses a library called Groovy CPS to run Pipeline
+scripts. While Pipeline uses the Groovy parser and compiler, unlike a
+regular Groovy environment it runs most of the program inside a special
+interpreter. This uses a continuation-passing style (CPS) transform to
+turn your code into a version that can save its current state to disk (a
+file called `+program.dat+`  inside your build directory) and continue
+running even after Jenkins has restarted. (You can get some more
+technical background on
+the https://github.com/jenkinsci/workflow-cps-plugin/blob/master/README.md[plugin page]
+and the https://github.com/cloudbees/groovy-cps/blob/master/README.md[library page].)
+
+While the CPS transform is usually transparent to users, there are
+limitations to what Groovy language constructs can be supported, and in
+some circumstances it can lead to counterintuitive behavior. https://issues.jenkins-ci.org/browse/JENKINS-31314[JENKINS-31314] makes the runtime try to detect the most common mistake: calling
+CPS-transformed code from non-CPS-transformed code. The following kinds
+of things are CPS-transformed:
+
+* Almost all of the Pipeline script you write (including in libraries).
+* Most Pipeline steps, including all those which take a block.
+
+The following kinds of things are _not_ CPS-transformed:
+
+* Compiled Java bytecode, including
+** the Java Platform
+** Jenkins core and plugins
+** the runtime for the Groovy language
+* Constructor bodies in your Pipeline script.
+* Any method in your Pipeline script marked with the `+@NonCPS+` 
+annotation.
+* A few Pipeline steps which take no block and act instantaneously, such
+as `+echo+` or `+properties+`.
+
+CPS-transformed code may call non-CPS-transformed code or other
+CPS-transformed code, and non-CPS-transformed code may call other
+non-CPS-transformed code, but it is impossible for non-CPS-transformed
+code to call CPS-transformed code. If you try, there will be a mismatch
+between the method that the CPS interpreter thought it was calling and
+the method that it actually got a return value from.
+
+== Common problems and solutions
+
+=== Use of Pipeline steps from `+@NonCPS+`
+
+Sometimes users will apply the `+@NonCPS+`  annotation to a method
+definition, which bypasses the CPS transform inside that method. This
```

Comment:
```suggestion
definition which bypasses the CPS transform inside that method. This
```

Reply:
I changed `, which bypasses` to `in order to bypass` in bb2cfe7.